torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp

```diff
@@ -157,32 +157,45 @@ static void setRequiresGradOnDiffGraph(Node* dnode) {
   // can  set df_input_vjps and DifferentiableGraphOp can set `requires_grad=`
   // properly
   auto go = dnode->g(attr::Subgraph)->outputs();
+  auto set_requires_grad = [](const TensorTypePtr& t, Value* val) -> bool {
+    if (t && t->requiresGrad().has_value()) {
+      GRAPH_DEBUG("setting type ", *t);
+      val->setType(t);
+      return true;
+    }
+    return false;
+  };
+
   for (size_t i = 0; i < go.size(); i++) {
     auto ty = go[i]->type()->cast<TensorType>();
     if (ty) {
       auto n = go[i]->node();
       auto dno = dnode->outputs().at(i);
-      auto dno_use0 = dno->uses().at(0);
-      GRAPH_DEBUG("found first user of ", i, " as ", *dno_use0.user);
-      if (n->kind() == prim::profile) {
-        GRAPH_DEBUG(
-            "setting output ", i, " to type ", *n->ty(attr::profiled_type));
-        go[i]->setType(n->ty(attr::profiled_type));
-      } else if (dno_use0.user->kind() == prim::profile) {
-        GRAPH_DEBUG(
-            "setting output ",
-            i,
-            " to type ",
-            *dno_use0.user->ty(attr::profiled_type));
-        go[i]->setType(dno_use0.user->ty(attr::profiled_type));
-      } else if (dno_use0.user->kind() == prim::DifferentiableGraph) {
-        Value* o =
-            dno_use0.user->g(attr::Subgraph)->inputs().at(dno_use0.offset);
-        auto nn = o->uses().at(0).user;
-        if (nn->kind() == prim::profile) {
-          GRAPH_DEBUG(
-              "setting output ", i, " to type ", *nn->ty(attr::profiled_type));
-          go[i]->setType(nn->ty(attr::profiled_type));
+      for (auto dno_use : dno->uses()) {
+        GRAPH_DEBUG("found user of ", i, " as ", *dno_use.user);
+        if (n->kind() == prim::profile) {
+          if (set_requires_grad(
+                  n->ty(attr::profiled_type)->expect<TensorType>(), go[i])) {
+            break;
+          }
+        } else if (dno_use.user->kind() == prim::profile) {
+          if (set_requires_grad(
+                  dno_use.user->ty(attr::profiled_type)->expect<TensorType>(),
+                  go[i])) {
+            break;
+          }
+        } else if (dno_use.user->kind() == prim::DifferentiableGraph) {
+          Value* o =
```

Comment:
oh god, this is really horrible. We really need to address this in a more systematic way as per your proposal about profile node propagation. If the next `DifferentiableGraph` will be guarded first before the current one, this logic will probably break. I guess we are kinda safe for now, since `diff_nodes` are in a linear order.